node/node.go

```diff
@@ -504,14 +484,147 @@ func NewNode(config *cfg.Config,
 	}
 
 	sw.SetAddrBook(addrBook)
+	return addrBook
+}
 
-	// run the profile server
+func runProfileServer(config *cfg.Config, logger log.Logger) {
 	profileHost := config.ProfListenAddress
 	if profileHost != "" {
 		go func() {
 			logger.Error("Profile server", "err", http.ListenAndServe(profileHost, nil))
 		}()
 	}
+}
+
+// NewNode returns a new, ready to go, Tendermint Node.
+func NewNode(config *cfg.Config,
+	privValidator types.PrivValidator,
+	nodeKey *p2p.NodeKey,
+	clientCreator proxy.ClientCreator,
+	genesisDocProvider GenesisDocProvider,
+	dbProvider DBProvider,
+	metricsProvider MetricsProvider,
+	logger log.Logger) (*Node, error) {
+
+	blockStore, stateDB, err := getStoresForNewNode(config, dbProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	state, genDoc, err := loadStateFromDBOrGenesisDoc(stateDB, genesisDocProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the proxyApp and establish connections to the ABCI app (consensus, mempool, query).
+	proxyApp, err := createAndStartProxyAppConns(clientCreator, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	// EventBus and IndexerService must be started before the handshake because
+	// we might need to index the txs of the replayed block as this might not have happened
+	// when the node stopped last time (i.e. the node stopped after it saved the block
+	// but before it indexed the txs, or, endblocker panicked)
+	eventBus, err := createAndStartEventBus(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	// Transaction indexing
+	indexerService, txIndexer, err := createAndStartIndexerService(config, dbProvider, eventBus, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the handshaker, which calls RequestInfo, sets the AppVersion on the state,
+	// and replays any blocks as necessary to sync tendermint with the app.
+	consensusLogger := logger.With("module", "consensus")
+	if err := doHandshakeForNewNode(stateDB, state, blockStore, genDoc, eventBus, proxyApp, consensusLogger); err != nil {
+		return nil, err
+	}
+
+	// Reload the state. It will have the Version.Consensus.App set by the
+	// Handshake, and may have other modifications as well (ie. depending on
+	// what happened during block replay).
+	state = sm.LoadState(stateDB)
+
+	// If an address is provided, listen on the socket for a connection from an
+	// external signing process.
+	privValidator, err = optionallyCreateAndStartPrivValidator(config, privValidator, logger)
```

Comment:
If we're going to use this `optionally` prefix (which I personally don't like because it brings uncertainty), we should document the expected behavior. Like `if the address is not provided, privValidator and err are both nil`. Note this won't prevent developers from making a mistake where they would assume `privValidator` can't ever be nil later in the code.

Reply:
In the case when both are `nil` we just asking for unexpected panics. Better to encode this in an explicit error which the caller can react on.